pkg/operation/botanist/resources.go

```diff
@@ -0,0 +1,60 @@
+// Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package botanist
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/gardener/gardener/extensions/pkg/controller"
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	"github.com/gardener/gardener/pkg/utils"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+const (
+	ManagedResourceName = "referenced-resources"
+)
+
+// DeployReferencedResources reads all referenced resources from the Garden cluster and writes a managed resource to the Seed cluster.
+func (b *Botanist) DeployReferencedResources(ctx context.Context) error {
+	// Read referenced objects into a slice of unstructured objects
+	var unstructuredObjs []*unstructured.Unstructured
+	for _, resourceRef := range b.Shoot.ResourceRefs {
+		// Read the resource from the Garden cluster
+		obj, err := utils.GetObjectByRef(ctx, b.K8sGardenClient.Client(), &resourceRef, b.Shoot.Info.Namespace)
+		if err != nil {
+			return err
+		}
+		if obj == nil {
+			return fmt.Errorf("object not found %v", resourceRef)
+		}
+
+		// Create an unstructured object and append it to the slice
+		unstructuredObj := &unstructured.Unstructured{Object: obj}
+		unstructuredObj.SetNamespace(b.Shoot.SeedNamespace)
+		unstructuredObjs = append(unstructuredObjs, &unstructured.Unstructured{Object: obj})
```

Comment:
Hmmm, can we maybe add some well-known prefix to the name of the resources we copy to the seed?
If the user specifies a resource name, we already use, e.g. Secret `gardener`, we will overwrite this resource, which can lead to problematic issues...

```suggestion
		unstructuredObj := &unstructured.Unstructured{Object: obj}
		unstructuredObj.SetName(SomeWellKnownPrefix + unstructuredObj.GetName())
		unstructuredObj.SetNamespace(b.Shoot.SeedNamespace)
		unstructuredObjs = append(unstructuredObjs, &unstructured.Unstructured{Object: obj})
```

Reply:
@tim-ebert @rfranzke I have some doubts. If we add a prefix when copying the resource, then the original reference in the Shoot would not be valid anymore (as it would use the unprefixed name), so any extension controller trying to use it would fail, unless they know to use the prefix. How exactly do you suggest to make this work?

Reply:
What about introducing a helper func like `GetObjectByResourceReference` which just takes a resource ref (which can be taken from the Shoot section of the Cluster resource) and returns the relating resource from the Seed? This function can then be used by the extension controllers and basically hides the prefix constant/naming mechanism from the caller, so they don't need the know the specifics...

Reply:
That could work, except that this helper func would have to return the object as `unstructured.Unstructured`, (or as `map[string]interface{}`) because we can't make any assumptions about its actual type. On the other hand, an extension controller could read the object as a strongly-typed struct, since they can make such assumptions. If were writing an extension controller, I would strongly prefer the strongly-typed option. If you are absolutely sure we need to do this, I would rather write a helper func `GetShootResourceReference(string)` that would return a correct `CrossVersionObjectReferece` by its name in the Shoot resource. I still have doubts however that all this would be useful enough to justify making the life of an extension controller developer harder in one way or another.

Reply:
> I still have doubts however that all this would be useful enough to justify making the life of an extension controller developer harder in one way or another.

The purpose of this change is not to be useful in any way, it's rather a requirement to avoid collision of names, right?

> If were writing an extension controller, I would strongly prefer the strongly-typed option.

I think, this should still be possible. You can make this helper func accept a resource ref and also a `runtime.Object`, so the caller can pass an `*corev1.Secret` for example and it gets filled by the library function (just like the controller-runtime client does it in `client.Get` and `client.List`). This way, it is strongly-typed and the caller just has to check if the resource ref points to a GroupKind that is supported (which they have to do either way) and then pass the ref and a pointer to the respective golang type to the helper func to retrieve the actual object. WDYT? Am I missing something?

Reply:
OK, did as suggested.